content/blog/2019-09-25-bril-c-backend.md

```diff
@@ -0,0 +1,92 @@
++++
+title = "A backend that translates Bril into C"
+extra.author = "Wen-Ding Li"
+extra.author_link = "https://www.cs.cornell.edu/~wdli/"
+extra.bio = """
+  [Wen-Ding Li](https://www.cs.cornell.edu/~wdli/) is a Ph.D. student at Cornell interested in high-performance code generation.
+"""
++++
+
+Bril is an educational compiler intermediate representation that is designed for this compiler course. While there is already a Bril interpreter written in
+typescript called `brili`, it would be interesting to have other backends so that we can compare the performance and the implementation complexity. In this first project,
+I built a backend that translates Bril into C language and then use GCC to compile and execute the program.
+
+Why choose C?
+---
+Translating to C provides some benefits such as portability, good performance, and easier integration with other C library and tools. C language is a widely used language in many embedded devices. We can also get native performance and leverage the GCC compiler's optimization. By translating to C, we can easily integrate it with other C library. Plus, because the Bril instructions (as it is now) can be mapped to C statements, we can also potentially use gdb as a debugger.
+Lastly, Translating to C is very common in the programming community and can be a fun project!
+
+Method
+---
+As of now, Bril only has one main function. For the sake of simplicity, I first collect all the name and type of variables used in the Bril program, and declare them at the top of the generated C code with the corresponding type, where I use `int64_t` for `int` and `int` for `bool`. Then, all the arithmetic, comparison, and logic instructions can be directed translate to corresponding C statement. For label and `jmp` instructions, I use `label` and `goto` in C. The instruction `br` can also be easily translated to statement compose of `if` and `goto`.
+
+In order to verify the correctness of this C backend, I create some valid handwritten tests to verify against the existing interpreter `brili`. The C backend successfully passes all the tests.
+
+Here we show a small example to demonstrate the translation.
+```
+main {
+  a: int = const 4;
+  b: int = const 4;
+  cmp: bool = ge a b;
+  jmp somewhere;
+  a: int = const 2;
+  somewhere:
+  c: int = add a b;
+  print c;
+  print cmp;
+}
+```
+The above Bril program can be translate to C as the following.
+```C
+#include <stdint.h>
+#include <stdio.h>
+#include <inttypes.h>
+int main(){
+int64_t a;
+int64_t b;
+int cmp;
+int64_t c;
+a = 4LL;
+b = 4LL;
+cmp = a >= b;
+goto somewhere;
+a = 2LL;
+somewhere:;
```

Comment:
Is this an extraneous `;` or intentional?

Reply:
Yes because we need to have a statement after a goto label in C. Extraneous `;` make it an empty statement and thus is valid. It is easier this way because in Bril we don't require a statement after label.